applications/FluidDynamicsApplication/custom_constitutive/newtonian_2d_law.cpp

```diff
@@ -79,6 +84,82 @@ void  Newtonian2DLaw::CalculateMaterialResponseCauchy(Parameters& rValues)
     }
 }
 
+void Newtonian2DLaw::CalculateDerivative(
+    Parameters& rParameterValues,
+    const Variable<Vector>& rFunctionVariable,
+    const Variable<double>& rDerivativeVariable,
+    Vector& rOutput)
+{
+    KRATOS_TRY
+
+    if (rFunctionVariable == CAUCHY_STRESS_VECTOR) {
+        // computes derivatives of CAUCHY_STRESS_VECTOR
+        // resize and clear the output variable
+        if (rOutput.size() != 3) {
+            rOutput.resize(3);
+        }
+        rOutput.clear();
+
+        if (rDerivativeVariable.IsComponent() && rDerivativeVariable.GetSourceVariable() == STRAIN_RATE_2D) {
+            // compute derivatives w.r.t. STRAIN_RATE
+            const double mu = this->GetEffectiveViscosity(rParameterValues);
+            const double volumetric_part_derivative = ((rDerivativeVariable.GetComponentIndex() == 0) + (rDerivativeVariable.GetComponentIndex() == 1)) / 3.0;
+
+            rOutput[0] = 2.0 * mu * ((rDerivativeVariable.GetComponentIndex() == 0) - volumetric_part_derivative);
+            rOutput[1] = 2.0 * mu * ((rDerivativeVariable.GetComponentIndex() == 1) - volumetric_part_derivative);
+            rOutput[2] = mu * (rDerivativeVariable.GetComponentIndex() == 2);
+
+        } else if (rDerivativeVariable == EFFECTIVE_VISCOSITY) {
+            // compute derivatives w.r.t. effective viscosity
+            const Vector& r_strain_rate = rParameterValues.GetStrainVector();
+            const double trace = r_strain_rate[0] + r_strain_rate[1];
+            const double volumetric_part = trace / 3.0;
+
+            rOutput[0] = 2.0 * (r_strain_rate[0] - volumetric_part);
+            rOutput[1] = 2.0 * (r_strain_rate[1] - volumetric_part);
+            rOutput[2] = r_strain_rate[2];
+        }
```

Comment:
Shouldn't we throw an error saying that `CAUCHY_STRESS_VECTOR` cannot be derived with respect to `rDerivatibleVariable.Name()`.

Reply:
Indeed, I'd call the base class in here.

Reply:
done :)